solc/CommandLineInterface.cpp

```diff
@@ -528,7 +528,7 @@ map<string, Json::Value> CommandLineInterface::parseAstFromInput()
 		}
 	}
 
-	m_fileReader.setSources(tmpSources);
+	m_fileReader.setSourceUnits(tmpSources);
```

Comment:
This is the place I talking about on the call. It's the only reason we needed `setSources()` at all. Looks like it do need the ability to set by source name after all.

But I think that we should just have a special constructor for this case and just create a fresh FileReader instance. It should not be possible to manipulate by source unit name once `FileReader` already exists.

Reply:
If we allow setting source units by their internal names, I don't see why this function would not be allowed as well.

Reply:
The problem is that `m_compiler` contains a reference to the file reader callback.

Reply:
> The problem is that `m_compiler` contains a reference to the file reader callback.

Yes, but file reader callback does not have to use this function. It can add stuff directly to `m_sources`.

> If we allow setting source units by their internal names, I don't see why this function would not be allowed as well.

I think we should not allow setting by source unit name. Only updating.

The reason is that if you insert a new source unit name, you have to ensure that it's relative to base path. It's error prone. I think you should have to go through the CLI -> source unit name translation because it's safer that way.

Reply:
Sure, but these are two distict ways we use the file reader. Maybe we could actually turn these into two different file readers.

Reply:
I really think we're constructing `FileReaders` to early, before we have all data for them. That's one thing I wanted to eventually refactor in the CLI (make it `optional` and initialize it once, only when I have all the data for it) and now there's the same thing with LSP where you're creating the adapter before you know the base path.

But anyway, I agree with https://github.com/ethereum/solidity/pull/12402#discussion_r767865777, that's a bit out of scope of this PR so we can do it some other time. For now it looks like `setSourceUnits()` unfortunately has to stay.